### PR TITLE
[#12421] Comments for Responses for Essay Question 

### LINKED ISSUE #12421: Comments for Responses for Essay Question
I am interested to work on this project: https://github.com/TEAMMATES/teammates/wiki/Comments-for-Responses. May I check if this issue is available? 

Currently, students can add comments to their responses when they answer MCQ or MSQ questions. it would be good to add a similar ability to the essay question type.

### PR DESCRIPTION
**Pull request by ellaella12**
Fixes #12421

Approach 1: 
- make  a pr to enable comments for each question type
- involves changing the flag to allow comment for a question type
- involves expanding set of question types that allows comment
- update test to ensure flag returns true

- when all question type allows comment, the 3 points above will be obsolete
- aka everything will be deleted 
- checks will no longer be done to check if question type allows for comment
(as all question allows comment)

Approach 2:
- 1 pr to enable comments for all question type
- removes flag, checker methods to check if  a question type allows comment
- the comment logic is done outside the question, so it is universal regardless of qn
- tests is done in one place to check comment logic

I chose approach 2 as it was easier, do let me know if you prefer breaking it down by question type.

- For now, submit responses for ALL  questions button, will not save comments
- only submit responses for question 1 button will save comment for qn 1
- this is the same for existing logic that allows comment for mcq/msq question
-  This pr is to ensure all question types works the same way as existing logic for mcq/msq
- I will make a seperate pr to address this
-  Or would you prefer I do it in this pr?

image
image
image
image
image

### CODE REVIEW

File: src/test/java/teammates/ui/webapi/GetFeedbackResponsesActionTest.java [hunk 1]
```diff
@@ -145,6 +145,30 @@ protected void testExecute_commentSubmission_shouldGetCommentsSuccessfully() thr
...
+    @Test
+    protected void testExecute_getTextQuestionType_shouldGetCommentsSuccessfully() throws Exception {
+        DataBundle dataBundle = loadDataBundle("/FeedbackResponseCommentCRUDTest.json");
+        removeAndRestoreDataBundle(dataBundle);
+
+        InstructorAttributes instructor1OfCourse1 = dataBundle.instructors.get("instructor1OfCourse1");
+        FeedbackQuestionAttributes qn2InSession1 = dataBundle.feedbackQuestions.get("qn2InSession1");
+        FeedbackResponseAttributes response1ForQ2 = dataBundle.feedbackResponses.get("response1ForQ2");
+        FeedbackResponseCommentAttributes comment1FromInstructor1Question2 =
+                dataBundle.feedbackResponseComments.get("comment1FromInstructor1Question2");
```
Comment (cedricongjh) [inline src/test/java/teammates/ui/webapi/GetFeedbackResponsesActionTest.java:157]: Given that the aim is for students to be able to comment on their responses, I think we should change this test to be a comment from a student instead?

**Review by cedricongjh — CHANGES_REQUESTED** [review]
Thank you for the PR! 

However, there appears to be errors when I try to submit a feedback response comment on non-MCQ and MRQ question types:

image

Do fix this and address the other comment regarding testing, thank you!

**Comment by ellaella12** [comment]
Hi, thank you for the review! may I check which qn would you be referring to that returns that error message? or how to replicate this? 
as i submitted comments for different qn types and got this
image

Thank you!

**Comment by ellaella12** [comment]
Hi, thank you for the review!

May I check how do i replicate the error message above? 
I tried submitting for different qn types and got this
image

**Comment by cedricongjh** [comment]
> Hi, thank you for the review!
> 
> May I check how do i replicate the error message above? I tried submitting for different qn types and got this image

HI @ellaella12, I tried again but did not run into the error, so no worries about this!

**Review by cedricongjh — CHANGES_REQUESTED** [review]
Hi @ellaella12, there's another thing to consider that I missed out, we need to add in the student's comment in the downloaded result as well:

Click download result:
image

The csv has `Giver comment` column for MCQ and MRQ type questions


Do add that in and re-request a review, thank you!

**Comment by ellaella12** [comment]
Hi, I have included the givers comment for all question types into the csv
for e.g, the images below, comment text are `cx` 
image
image
image


Thank you!

**Review by cedricongjh — COMMENTED** [review]
LGTM, thank you for working on this project!

**Comment by kevin9foong** [comment]
@cedricongjh, not sure if this is ok to be merged

**Comment by cedricongjh** [comment]
> @cedricongjh, not sure if this is ok to be merged

It needs a final review from @zhaojj2209, so not yet!

**Comment by zhaojj2209** [comment]
Apologies for the delay, will try to get to this PR within the next few days.

**Comment by ellaella12** [comment]
sure, no rush, thank you!

**Comment by zhaojj2209** [comment]
Great work so far!

One bug I've found: for distribute points questions and team contribution questions, the component state is not being updated correctly and only updates on refresh. See below video where the comment text editor is still active after saving the comment, and attempting to submit again causes an error. For deletion, the comment does not disappear immediately after deleting, and an error will pop up when attempting to submit again.

https://github.com/TEAMMATES/teammates/assets/54243224/a7e03a4f-d600-43c0-a0d6-584302ab782b

For reference, the correct behaviour is as follows:

https://github.com/TEAMMATES/teammates/assets/54243224/38c5ecda-da21-4144-bbaa-cb9de1690ea0

As for your question regarding the saving of comments when saving all questions, I agree that it should be in a separate PR, with a separate issue to track it.

**Comment by ellaella12** [comment]
Hi thank you for the review! I have fixed the bug and its able to delete and submit in real time now.
I have a few questions:

1. the Bug is fixed by updating the model before emitting an event to the parent, like how 2 way binding is done in angular docs https://angular.io/guide/two-way-binding. Originally, it doesn't update the model but just sends the updated values to the parent. I am unsure why it works and why do we have to update the model before emitting event to parent?

Is it because the stale child model override the parent model?

2. From a quick check, it seems the `CommentEditFormModel.isUsingCustomVisiblities` is not used as the comment follows the question's visibility. Furthermore this visibilities is set to empty after a comment is edited. Is it used somewhere?

3. After switching to master branch and running e2e tests, it is failing at the same test case as this pr.

**Comment by zhaojj2209** [comment]
1. > Is it because the stale child model override the parent model?

From what I know that is indeed the case. Thanks for helping to fix this!

2. `isUsingCustomVisibilities` is used in `comment-edit-form.component.ts`.

3. We will help to monitor the E2E test runs, sometimes tests take multiple runs to pass.

**Comment by ellaella12** [comment]
Hi, sorry about that I have fixed it, it should work now. Thank you!

[TEAMMATES - Online Peer Feedback_Evaluation System for Student Team Projects (1).webm](https://github.com/TEAMMATES/teammates/assets/120079323/8ce463cd-da11-422b-a47a-83dc85e35a89)

**Comment by ellaella12** [comment]
Hi sorry about that, I have made a fix, it's working now
[TEAMMATES - Online Peer Feedback_Evaluation System for Student Team Projects (1).webm](https://github.com/TEAMMATES/teammates/assets/120079323/f1ed8149-c63a-4c19-950e-fdcb42b3ca06)

**Comment by zhaojj2209** [comment]
Thanks for the fix, great work so far!

One last issue I've managed to uncover: after submitting a feedback response to a distribute points (recipients) question, reloading the page and attempting to submit it again without making any changes will give the following error:

Screenshot 2023-06-24 at 22 17 14

The issue doesn't appear if you make any changes to the response before submitting the response again, and the issue is not present for any other question type.

Let me know if you need more clarifications on when the bug occurs.

This should be the last bug left (crosses fingers), after this the PR should be good to go 💪

**Comment by ellaella12** [comment]
Hi, thank you for the review! Sorry for the numerous bugs. have made a fix for submission without changing answer, by changing `QuestionConstraintComponent` to check is valid only in `ngOnInit()` when both the `recipientSubmissionForms` and `questionDetails` has initialized. 
Originally, the check is valid is called right after `recipientSubmissionForms`, so default `questionDetails` are used = wrong validity check. Is this fix ok?
[TEAMMATES - Online Peer Feedback_Evaluation System for Student Team Projects (2).webm](https://github.com/TEAMMATES/teammates/assets/120079323/605c3aa7-805c-4c14-8694-0f871c7ef264)


For this ```Edit: My bad, found one more bug 🙇‍♀️ The response comments are not being submitted when clicking the "Submit Responses to All Questions" button, do look into it.```
The comment dated May 28 ``As for your question regarding the saving of comments when saving all questions, I agree that it should be in a separate PR, with a separate issue to track it.```
Or should the change be implemented in this pr?

**Review by zhaojj2209 — APPROVED** [review]
Thank you so much for the work done!

@wkurniawan07 Now that we are allowing feedback response participants to add comments for every question type, should we evaluate whether we should do the same for instructor comments as well? Currently instructors are allowed to add comments on all question types except team contribution.

**Comment by damithc** [comment]
> @wkurniawan07 Now that we are allowing feedback response participants to add comments for every question type, should we evaluate whether we should do the same for instructor comments as well? Currently instructors are allowed to add comments on all question types except team contribution.

@zhaojj2209 Perhaps it is better to leave out team contribution questions from this feature. The values shown to instructors could be different from those shown to students (due to normalizing etc.), so the comments instructors enter could be misleading to students.

**Review by cedricongjh — APPROVED** [review]
LGTM, great work on this!

**Pull request merged**